In [2]:
import numpy as np
import pandas as pd
df = pd.read_csv('playstore_apps.csv')

In [2]:
df

App             Category  \
0         Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                    Coloring book moana       ART_AND_DESIGN   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                  Sketch - Draw & Paint       ART_AND_DESIGN   
4                  Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                  ...                  ...   
10836                                   Sya9a Maroc - FR               FAMILY   
10837                   Fr. Mike Schmitz Audio Teachings               FAMILY   
10838                             Parkinson Exercices FR              MEDICAL   
10839                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       Rating   Reviews                Size    Installs  Type  Price  \
0         4.1     159.0                 19M     10000.0  Free    0.0   
1         3.9     967.0                 14M    500000.0  Free    0.0   
2         4.7   87510.0                8.7M   5000000.0  Free    0.0   
3         4.5  215644.0                 25M  50000000.0  Free    0.0   
4         4.3     967.0                2.8M    100000.0  Free    0.0   
...       ...       ...                 ...         ...   ...    ...   
10836     4.5      38.0                 53M      5000.0  Free    0.0   
10837     5.0       4.0                3.6M       100.0  Free    0.0   
10838     NaN       3.0                9.5M      1000.0  Free    0.0   
10839     4.5     114.0  Varies with device      1000.0  Free    0.0   
10840     4.5  398307.0                 19M  10000000.0  Free    0.0   

      Content Rating                     Genres Last Updated  \
0           Everyone               Art & Design   07-01-2018   
1           Everyone  Art & Design;Pretend Play   15-01-2018   
2           Everyone               Art & Design   01-08-2018   
3               Teen               Art & Design   08-06-2018   
4           Everyone    Art & Design;Creativity   20-06-2018   
...              ...                        ...          ...   
10836       Everyone                  Education   25-07-2017   
10837       Everyone                  Education   06-07-2018   
10838       Everyone                    Medical   20-01-2017   
10839     Mature 17+          Books & Reference   19-01-2015   
10840       Everyone                  Lifestyle   25-07-2018   

              Current Ver         Android Ver  
0                   1.0.0        4.0.3 and up  
1                   2.0.0        4.0.3 and up  
2                   1.2.4        4.0.3 and up  
3      Varies with device          4.2 and up  
4                     1.1          4.4 and up  
...                   ...                 ...  
10836                1.48          4.1 and up  
10837                   1          4.1 and up  
10838                   1          2.2 and up  
10839  Varies with device  Varies with device  
10840  Varies with device  Varies with device  

[10841 rows x 13 columns]

In [3]:
df.isnull().sum().sum()

1491

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          9367 non-null   float64
 3   Reviews         10840 non-null  float64
 4   Size            10841 non-null  object 
 5   Installs        10840 non-null  float64
 6   Type            10840 non-null  object 
 7   Price           10840 non-null  float64
 8   Content Rating  10840 non-null  object 
 9   Genres          10841 non-null  object 
 10  Last Updated    10840 non-null  object 
 11  Current Ver     10833 non-null  object 
 12  Android Ver     10838 non-null  object 
dtypes: float64(4), object(9)
memory usage: 1.1+ MB


In total 1491 null values are there and majority of them are in Rating column and there are few missing values in other features.

# Dealing With missing values

In [5]:
df[df['Reviews'].isnull()==True]

App Category  Rating  Reviews  \
10472  Life Made WI-Fi Touchscreen Photo Frame      1.9    19.0      NaN   

         Size  Installs Type  Price Content Rating             Genres  \
10472  1,000+       NaN    0    NaN            NaN  February 11, 2018   

      Last Updated Current Ver Android Ver  
10472          NaN  4.0 and up         NaN

In [6]:
df['Category'].value_counts()

FAMILY                 1972
GAME                   1144
TOOLS                   843
MEDICAL                 463
BUSINESS                460
PRODUCTIVITY            424
PERSONALIZATION         392
COMMUNICATION           387
SPORTS                  384
LIFESTYLE               382
FINANCE                 366
HEALTH_AND_FITNESS      341
PHOTOGRAPHY             335
SOCIAL                  295
NEWS_AND_MAGAZINES      283
SHOPPING                260
TRAVEL_AND_LOCAL        258
DATING                  234
BOOKS_AND_REFERENCE     231
VIDEO_PLAYERS           175
EDUCATION               156
ENTERTAINMENT           149
MAPS_AND_NAVIGATION     137
FOOD_AND_DRINK          127
HOUSE_AND_HOME           88
LIBRARIES_AND_DEMO       85
AUTO_AND_VEHICLES        85
WEATHER                  82
ART_AND_DESIGN           65
EVENTS                   64
COMICS                   60
PARENTING                60
BEAUTY                   53
1.9                       1
Name: Category, dtype: int64

So here I will drop this entire record as it has 6 null values and it's Category is wrong I believe as it is the integer and different from all other categories

In [7]:
df.drop(df[df['Category']=='1.9'].index,inplace=True)

In [8]:
# Type

In [9]:
df['Type'].value_counts()

Free    10039
Paid      800
Name: Type, dtype: int64

In [10]:
df[df['Type'].isnull()==True]

App Category  Rating  Reviews                Size  \
9148  Command & Conquer: Rivals   FAMILY     NaN      0.0  Varies with device   

      Installs Type  Price Content Rating    Genres Last Updated  \
9148       0.0  NaN    0.0   Everyone 10+  Strategy   28-06-2018   

             Current Ver         Android Ver  
9148  Varies with device  Varies with device

So here I will drop this entire record as it has 2 null values and it's installs are zero as well.

In [11]:
df.drop(df[df['Type'].isnull()==True].index,inplace=True)

In [12]:
# Current Ver

In [13]:
df['Current Ver'].value_counts()

Varies with device    1458
1                      842
1.1                    276
1.2                    185
2                      165
                      ... 
2.5_20                   1
V1.0.1                   1
1.2.1.0                  1
8.0.1.0                  1
1.5.341                  1
Name: Current Ver, Length: 2783, dtype: int64

In [14]:
df.drop(df[df['Current Ver'].isnull()==True].index,inplace=True)

In [15]:
df

App             Category  \
0         Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                    Coloring book moana       ART_AND_DESIGN   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                  Sketch - Draw & Paint       ART_AND_DESIGN   
4                  Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                  ...                  ...   
10836                                   Sya9a Maroc - FR               FAMILY   
10837                   Fr. Mike Schmitz Audio Teachings               FAMILY   
10838                             Parkinson Exercices FR              MEDICAL   
10839                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       Rating   Reviews                Size    Installs  Type  Price  \
0         4.1     159.0                 19M     10000.0  Free    0.0   
1         3.9     967.0                 14M    500000.0  Free    0.0   
2         4.7   87510.0                8.7M   5000000.0  Free    0.0   
3         4.5  215644.0                 25M  50000000.0  Free    0.0   
4         4.3     967.0                2.8M    100000.0  Free    0.0   
...       ...       ...                 ...         ...   ...    ...   
10836     4.5      38.0                 53M      5000.0  Free    0.0   
10837     5.0       4.0                3.6M       100.0  Free    0.0   
10838     NaN       3.0                9.5M      1000.0  Free    0.0   
10839     4.5     114.0  Varies with device      1000.0  Free    0.0   
10840     4.5  398307.0                 19M  10000000.0  Free    0.0   

      Content Rating                     Genres Last Updated  \
0           Everyone               Art & Design   07-01-2018   
1           Everyone  Art & Design;Pretend Play   15-01-2018   
2           Everyone               Art & Design   01-08-2018   
3               Teen               Art & Design   08-06-2018   
4           Everyone    Art & Design;Creativity   20-06-2018   
...              ...                        ...          ...   
10836       Everyone                  Education   25-07-2017   
10837       Everyone                  Education   06-07-2018   
10838       Everyone                    Medical   20-01-2017   
10839     Mature 17+          Books & Reference   19-01-2015   
10840       Everyone                  Lifestyle   25-07-2018   

              Current Ver         Android Ver  
0                   1.0.0        4.0.3 and up  
1                   2.0.0        4.0.3 and up  
2                   1.2.4        4.0.3 and up  
3      Varies with device          4.2 and up  
4                     1.1          4.4 and up  
...                   ...                 ...  
10836                1.48          4.1 and up  
10837                   1          4.1 and up  
10838                   1          2.2 and up  
10839  Varies with device  Varies with device  
10840  Varies with device  Varies with device  

[10831 rows x 13 columns]

In [16]:
# Android Ver

In [17]:
df['Android Ver'].value_counts()

4.1 and up            2451
4.0.3 and up          1501
4.0 and up            1373
Varies with device    1361
4.4 and up             980
2.3 and up             651
5.0 and up             601
4.2 and up             393
2.3.3 and up           281
2.2 and up             244
4.3 and up             243
3.0 and up             240
2.1 and up             134
1.6 and up             115
6.0 and up              60
7.0 and up              42
3.2 and up              36
2.0 and up              31
5.1 and up              24
1.5 and up              19
4.4W and up             12
3.1 and up              10
2.0.1 and up             7
8.0 and up               6
7.1 and up               3
4.0.3 - 7.1.1            2
5.0 - 8.0                2
1.0 and up               2
5.0 - 6.0                1
7.0 - 7.1.1              1
4.1 - 7.1.1              1
2.2 - 7.1.1              1
5.0 - 7.1.1              1
Name: Android Ver, dtype: int64

In [18]:
df[df['Android Ver'].isnull()==True]

App         Category  Rating  Reviews  Size  \
4453  [substratum] Vacuum: P  PERSONALIZATION     4.4    230.0   11M   
4490    Pi Dark [substratum]  PERSONALIZATION     4.5    189.0  2.1M   

      Installs  Type  Price Content Rating           Genres Last Updated  \
4453    1000.0  Paid   1.49       Everyone  Personalization   20-07-2018   
4490   10000.0  Free   0.00       Everyone  Personalization   27-03-2018   

     Current Ver Android Ver  
4453         4.4         NaN  
4490         1.1         NaN

In [19]:
df.drop(df[df['Android Ver'].isnull()==True].index,inplace=True)

In [20]:
df

App             Category  \
0         Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                    Coloring book moana       ART_AND_DESIGN   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                  Sketch - Draw & Paint       ART_AND_DESIGN   
4                  Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                  ...                  ...   
10836                                   Sya9a Maroc - FR               FAMILY   
10837                   Fr. Mike Schmitz Audio Teachings               FAMILY   
10838                             Parkinson Exercices FR              MEDICAL   
10839                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       Rating   Reviews                Size    Installs  Type  Price  \
0         4.1     159.0                 19M     10000.0  Free    0.0   
1         3.9     967.0                 14M    500000.0  Free    0.0   
2         4.7   87510.0                8.7M   5000000.0  Free    0.0   
3         4.5  215644.0                 25M  50000000.0  Free    0.0   
4         4.3     967.0                2.8M    100000.0  Free    0.0   
...       ...       ...                 ...         ...   ...    ...   
10836     4.5      38.0                 53M      5000.0  Free    0.0   
10837     5.0       4.0                3.6M       100.0  Free    0.0   
10838     NaN       3.0                9.5M      1000.0  Free    0.0   
10839     4.5     114.0  Varies with device      1000.0  Free    0.0   
10840     4.5  398307.0                 19M  10000000.0  Free    0.0   

      Content Rating                     Genres Last Updated  \
0           Everyone               Art & Design   07-01-2018   
1           Everyone  Art & Design;Pretend Play   15-01-2018   
2           Everyone               Art & Design   01-08-2018   
3               Teen               Art & Design   08-06-2018   
4           Everyone    Art & Design;Creativity   20-06-2018   
...              ...                        ...          ...   
10836       Everyone                  Education   25-07-2017   
10837       Everyone                  Education   06-07-2018   
10838       Everyone                    Medical   20-01-2017   
10839     Mature 17+          Books & Reference   19-01-2015   
10840       Everyone                  Lifestyle   25-07-2018   

              Current Ver         Android Ver  
0                   1.0.0        4.0.3 and up  
1                   2.0.0        4.0.3 and up  
2                   1.2.4        4.0.3 and up  
3      Varies with device          4.2 and up  
4                     1.1          4.4 and up  
...                   ...                 ...  
10836                1.48          4.1 and up  
10837                   1          4.1 and up  
10838                   1          2.2 and up  
10839  Varies with device  Varies with device  
10840  Varies with device  Varies with device  

[10829 rows x 13 columns]

## Dealing with missing values in Rating Feature

In [21]:
from sklearn.impute import KNNImputer 

In [22]:
imputer = KNNImputer(n_neighbors=12)
df['Rating']=imputer.fit_transform(df[['Rating']])

In [23]:
df.isna().sum()

App               0
Category          0
Rating            0
Reviews           0
Size              0
Installs          0
Type              0
Price             0
Content Rating    0
Genres            0
Last Updated      0
Current Ver       0
Android Ver       0
dtype: int64

## Dealing with duplicated values

In [24]:
df.duplicated('App')

0        False
1        False
2        False
3        False
4        False
         ...  
10836    False
10837    False
10838    False
10839    False
10840    False
Length: 10829, dtype: bool

In [25]:
df[df.duplicated('App')==True]

App       Category  Rating   Reviews  \
229          Quick PDF Scanner + OCR FREE       BUSINESS     4.2   80805.0   
236                                   Box       BUSINESS     4.2  159872.0   
239                    Google My Business       BUSINESS     4.4   70991.0   
256                   ZOOM Cloud Meetings       BUSINESS     4.4   31614.0   
261             join.me - Simple Meetings       BUSINESS     4.0    6989.0   
...                                   ...            ...     ...       ...   
10715                  FarmersOnly Dating         DATING     3.0    1145.0   
10720  Firefox Focus: The privacy browser  COMMUNICATION     4.4   36981.0   
10730                         FP Notebook        MEDICAL     4.5     410.0   
10753      Slickdeals: Coupons & Shopping       SHOPPING     4.5   33599.0   
10768                                AAFP        MEDICAL     3.8      63.0   

                     Size    Installs  Type  Price Content Rating  \
229    Varies with device   5000000.0  Free    0.0       Everyone   
236    Varies with device  10000000.0  Free    0.0       Everyone   
239    Varies with device   5000000.0  Free    0.0       Everyone   
256                   37M  10000000.0  Free    0.0       Everyone   
261    Varies with device   1000000.0  Free    0.0       Everyone   
...                   ...         ...   ...    ...            ...   
10715                1.4M    100000.0  Free    0.0     Mature 17+   
10720                4.0M   1000000.0  Free    0.0       Everyone   
10730                 60M     50000.0  Free    0.0       Everyone   
10753                 12M   1000000.0  Free    0.0       Everyone   
10768                 24M     10000.0  Free    0.0       Everyone   

              Genres Last Updated         Current Ver         Android Ver  
229         Business   26-02-2018  Varies with device        4.0.3 and up  
236         Business   31-07-2018  Varies with device  Varies with device  
239         Business   24-07-2018    2.19.0.204537701          4.4 and up  
256         Business   20-07-2018      4.1.28165.0716          4.0 and up  
261         Business   16-07-2018           4.3.0.508          4.4 and up  
...              ...          ...                 ...                 ...  
10715         Dating   25-02-2016                 2.2          4.0 and up  
10720  Communication   06-07-2018                 5.2          5.0 and up  
10730        Medical   24-03-2018           2.1.0.372          4.4 and up  
10753       Shopping   30-07-2018                 3.9          4.4 and up  
10768        Medical   22-06-2018               2.3.1          5.0 and up  

[1181 rows x 13 columns]

In [26]:
df[df['App']=='Quick PDF Scanner + OCR FREE']

App  Category  Rating  Reviews  \
222  Quick PDF Scanner + OCR FREE  BUSINESS     4.2  80805.0   
229  Quick PDF Scanner + OCR FREE  BUSINESS     4.2  80805.0   
285  Quick PDF Scanner + OCR FREE  BUSINESS     4.2  80804.0   

                   Size   Installs  Type  Price Content Rating    Genres  \
222  Varies with device  5000000.0  Free    0.0       Everyone  Business   
229  Varies with device  5000000.0  Free    0.0       Everyone  Business   
285  Varies with device  5000000.0  Free    0.0       Everyone  Business   

    Last Updated         Current Ver   Android Ver  
222   26-02-2018  Varies with device  4.0.3 and up  
229   26-02-2018  Varies with device  4.0.3 and up  
285   26-02-2018  Varies with device  4.0.3 and up

In [27]:
df.drop_duplicates('App',keep='first',inplace = True)

In [30]:
df

App             Category  \
0         Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                    Coloring book moana       ART_AND_DESIGN   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                  Sketch - Draw & Paint       ART_AND_DESIGN   
4                  Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                  ...                  ...   
10836                                   Sya9a Maroc - FR               FAMILY   
10837                   Fr. Mike Schmitz Audio Teachings               FAMILY   
10838                             Parkinson Exercices FR              MEDICAL   
10839                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

         Rating   Reviews                Size    Installs  Type  Price  \
0      4.100000     159.0                 19M     10000.0  Free    0.0   
1      3.900000     967.0                 14M    500000.0  Free    0.0   
2      4.700000   87510.0                8.7M   5000000.0  Free    0.0   
3      4.500000  215644.0                 25M  50000000.0  Free    0.0   
4      4.300000     967.0                2.8M    100000.0  Free    0.0   
...         ...       ...                 ...         ...   ...    ...   
10836  4.500000      38.0                 53M      5000.0  Free    0.0   
10837  5.000000       4.0                3.6M       100.0  Free    0.0   
10838  4.191838       3.0                9.5M      1000.0  Free    0.0   
10839  4.500000     114.0  Varies with device      1000.0  Free    0.0   
10840  4.500000  398307.0                 19M  10000000.0  Free    0.0   

      Content Rating                     Genres Last Updated  \
0           Everyone               Art & Design   07-01-2018   
1           Everyone  Art & Design;Pretend Play   15-01-2018   
2           Everyone               Art & Design   01-08-2018   
3               Teen               Art & Design   08-06-2018   
4           Everyone    Art & Design;Creativity   20-06-2018   
...              ...                        ...          ...   
10836       Everyone                  Education   25-07-2017   
10837       Everyone                  Education   06-07-2018   
10838       Everyone                    Medical   20-01-2017   
10839     Mature 17+          Books & Reference   19-01-2015   
10840       Everyone                  Lifestyle   25-07-2018   

              Current Ver         Android Ver  
0                   1.0.0        4.0.3 and up  
1                   2.0.0        4.0.3 and up  
2                   1.2.4        4.0.3 and up  
3      Varies with device          4.2 and up  
4                     1.1          4.4 and up  
...                   ...                 ...  
10836                1.48          4.1 and up  
10837                   1          4.1 and up  
10838                   1          2.2 and up  
10839  Varies with device  Varies with device  
10840  Varies with device  Varies with device  

[9648 rows x 13 columns]

In [31]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9648 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             9648 non-null   object 
 1   Category        9648 non-null   object 
 2   Rating          9648 non-null   float64
 3   Reviews         9648 non-null   float64
 4   Size            9648 non-null   object 
 5   Installs        9648 non-null   float64
 6   Type            9648 non-null   object 
 7   Price           9648 non-null   float64
 8   Content Rating  9648 non-null   object 
 9   Genres          9648 non-null   object 
 10  Last Updated    9648 non-null   object 
 11  Current Ver     9648 non-null   object 
 12  Android Ver     9648 non-null   object 
dtypes: float64(4), object(9)
memory usage: 1.0+ MB


In [32]:
df.to_csv('final_playstore_apps.csv',index=False)

In [4]:
df['Category'].value_counts()

FAMILY                 1972
GAME                   1144
TOOLS                   843
MEDICAL                 463
BUSINESS                460
PRODUCTIVITY            424
PERSONALIZATION         392
COMMUNICATION           387
SPORTS                  384
LIFESTYLE               382
FINANCE                 366
HEALTH_AND_FITNESS      341
PHOTOGRAPHY             335
SOCIAL                  295
NEWS_AND_MAGAZINES      283
SHOPPING                260
TRAVEL_AND_LOCAL        258
DATING                  234
BOOKS_AND_REFERENCE     231
VIDEO_PLAYERS           175
EDUCATION               156
ENTERTAINMENT           149
MAPS_AND_NAVIGATION     137
FOOD_AND_DRINK          127
HOUSE_AND_HOME           88
AUTO_AND_VEHICLES        85
LIBRARIES_AND_DEMO       85
WEATHER                  82
ART_AND_DESIGN           65
EVENTS                   64
PARENTING                60
COMICS                   60
BEAUTY                   53
1.9                       1
Name: Category, dtype: int64

In [5]:
df.columns

Index(['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type',
       'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver',
       'Android Ver'],
      dtype='object')